File: uglynumber.java
//Time complexity-O(n)
//Space complexity-O(n)
//Ran on leetcode
class Solution {
    public int nthUglyNumber(int n) {
        ArrayList<Integer> list = new ArrayList<Integer>();
        list.add(1);
        int i=0,j=0,k=0;
        while(list.size()<n){
            list.add(Math.min(Math.min(list.get(i)*2,list.get(j)*3),list.get(k)*5));
            if(list.get(list.size() - 1)==list.get(i)*2) ++i;
            if(list.get(list.size() - 1)==list.get(j)*3) ++j;
            if(list.get(list.size() - 1)==list.get(k)*5) ++k;
        }
        return list.get(list.size() - 1);
    }
}

File: UglyNumber.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        '''
        Time: O(nlogn)
        Space: O(n)
        '''
        import heapq
        x = set([1])
        hp = [1]
        max1 = 1
        j=1
        z = [2,3,5]
        while(j<=n):
            h  = heapq.heappop(hp)
            j+=1
            if(h>max1):
                max1 = h
            for i in z:
                if(not i*h in x):
                    heapq.heappush(hp,i*h)
                    x.add(i*h)
        
        return max1
